### ISSUE TITLE
Embeds from remote servers running 4.3 not working

### ISSUE THREAD

**Issue by joenepraat** [mastodon/mastodon#32408]
### Steps to reproduce the problem

1. Try to get the embed code for a post from a user on a server that's running Mastodon version 4.3.
2. The embed modal is almost empty and the following error is generated in Firefox web tools: `Content-Security-Policy: The page’s settings blocked an inline style (style-src-elem) from being applied because it violates the following directive: “style-src 'unsafe-inline' 'self' https://todon.nl 'nonce-random_22_character_string=='”`
3. Embeds from users hosted on a server below v4.3 are working fine. Also note that embeds from local posts are also working fine, despite running v4.3.

### Expected behaviour

Receiving the embed code

### Actual behaviour

Not receiving the embed code for remote server running v4.3

### Detailed description

_**I tested this also on mastodon.social and received the same results.**_

### Mastodon instance

todon.nl

### Mastodon version

v4.3.0

### Browser name and version

Firefox 131.0

### Operating system

Ubuntu Linux 24.04 LTS

**Comment by ClearlyClaire** [comment]
To add some background to this issue:
- for remote posts, the remote server is responsible for providing the embed code
- because we can't blindly trust the remote embed code, we sanitize it, only allowing a handful of HTML tags (such as `iframe`, `audio` or `video`) and attributes
- until 4.3.0, Mastodon embeds worked with an `iframe` and some auxiliary javascript to resize that embed, but now they rely on a `blockquote` with some styling and data attributes for a placeholder, plus a script that loads everything else; our sanitization strips all of that

This is a bit of a tough issue, because:
- the new embeds rely on script execution to be useful
- we don't want to present users with arbitrary HTML we haven't vetted

**Comment by ClearlyClaire** [comment]
We will likely disable the menu item (see <https://github.com/mastodon/mastodon/pull/32578>) for remote posts for the time being, until we find a solution to this issue.

**Comment by renchap** [comment]
A possible solution we are discussing is not loading the *remote* embed code, but always offer the user with embed code from their current instance, which will render the *current instance* view of the remote content.

This would avoid having to trust other instances, lot 1 JS file for every instance you want to embed from, and also work with non-Mastodon remote content (which has never worked).

**Comment by ThisIsMissEm** [comment]
As an example, this post: https://mastodon.social/@_elena/113351163816987429

When on a remote server to mastodon.social and you try to create an embed, the returned json for the embed is:

```
{
  "type": "rich",
  "version": "1.0",
  "author_name": "Elena Rossini ⁂",
  "author_url": "https://mastodon.social/@_elena",
  "provider_name": "mastodon.social",
  "provider_url": "https://mastodon.social/",
  "cache_age": 86400,
  "html": "     Post by @_elena@mastodon.social   View on Mastodon     ",
  "width": 400,
  "height": null
}
```

I think going with that "always embed from the current instance" is probably the way to go, though it would be nice to use the remote embed if possible.

**Comment by flancian** [comment]
Hi there! I've gotten here because I develop a social knowledge graph (anagora.org) that embeds Mastodon posts when they are found in user contributions; embeds worked fine until 4.2 and have been broken for posts served by 4.3 since that was released. Only the placeholder ever renders, the embed.js that you send with the blockquote never runs it would seem.

I'm unsure if I should open a new bug, as this seems related to the change you mention (from iframe to blockquote + js) above; please let me know what you think. I think it's worth tracking regardless because I would guess that other software apart from Mastodon itself and my nice knowledge graph also broke with 4.3?

**Comment by ClearlyClaire** [comment]
I think it'd be worth opening a new bug, as this is not the same issue. In either case, make sure the `embed.js` is not blocked by your Content-Security-Policy. It is reasonable to block such scripts, especially if you are including embeds from multiple un-vetted sources on the fly, but if you do so, you could use your own copy of `embed.js` allowing the expected domains.

**Comment by flancian** [comment]
Thank you @ClearlyClaire, I have opened https://github.com/mastodon/mastodon/issues/33049 to track my issue and included some observations on which parts of embed.js seem to be failing on what seems to me a standard setup/use case.

**Comment by ThisIsMissEm** [comment]
@ClearlyClaire fwiw, I saw the CSP warning when on m.s and getting the embed code for an m.s status

**Comment by ClearlyClaire** [comment]
Yes, #34776 is indeed not exactly the same issue, but it's in the same ballpark. There's not much we can do about that. Basically, since we create an `iframe` dynamically, it has the same context, including CSP as the page creating it, while the embed has inline styles for the fallback placeholder.
